src/silo/preprocessing/preprocessor.cpp

```diff
@@ -0,0 +1,515 @@
+               silo::preprocessing::FASTA_EXTENSION
+            )
+         );
+         silo::FastaReader fasta_reader(preprocessing_config.getGeneFilenameNoExtension(seq_name)
+                                           .replace_extension(silo::preprocessing::FASTA_EXTENSION)
+         );
+
+         auto return_code = preprocessing_db.query(fmt::format(
+
+            R"-(
+create or replace view gene_{0} as
+select key, sequence,
+partitioned_metadata.partition_id as partition_id
+{2}
+from raw_gene_{0} raw, partitioned_metadata
+where raw.key = partitioned_metadata.{1};
+)-",
+            seq_name,
+            database_config.schema.primary_key,
+            order_by_select
+         ));
+      }
+   }
+
+   preprocessing::Partitions partition_descriptor = preprocessing_db.getPartitionDescriptor();
+
+   SPDLOG_INFO("preprocessing - building database");
+
+   return buildDatabase(partition_descriptor, reference_genomes, order_by_clause, alias_key);
+}
+
+void Preprocessor::buildPartitioningTable() {
+   if (database_config.schema.partition_by.has_value()) {
+      SPDLOG_INFO("preprocessing - calculating partitions");
+
+      auto return_code = preprocessing_db.query(fmt::format(
+         R"-(
+create
+or replace table partition_keys as
+select row_number() over () - 1 as id, partition_key, count
+from (SELECT {} as partition_key, COUNT(*) as count
+      FROM metadata_table
+      GROUP BY partition_key
+      ORDER BY partition_key);
+)-",
+         database_config.schema.partition_by.value()
+      ));
+
+      if (return_code->HasError()) {
+         throw PreprocessingException(
+            "Error in the execution of the duckdb statement for partition key table "
+            "generation: " +
+            return_code->GetError()
+         );
+      } else {
+         SPDLOG_TRACE("Executed statement for partition key table generation.");
+         SPDLOG_TRACE(return_code->ToString());
+      }
+
+      return_code = preprocessing_db.query(
+         R"-(
+create or replace table partitioning as
+with recursive
+          allowed_count(allowed_count) as (select sum(count) / 32 from
+partition_keys),
+          grouped_partition_keys(from_id, to_id, count) as
+              (select id, id, count
+               from partition_keys
+               where id = 0
+               union all
+               select case when l1.count <= allowed_count then l1.from_id else l2.id end,
+                      l2.id,
+                      case when l1.count <= allowed_count
+                           then l1.count + l2.count
+                           else l2.count end
+               from grouped_partition_keys l1,
+                    partition_keys l2,
+                    allowed_count
+where l1.to_id + 1 = l2.id)
+select row_number() over () - 1 as partition_id, from_id, to_id, count
+from (select from_id, max(to_id) as to_id, max(count) as count
+      from grouped_partition_keys
+      group by from_id)
+)-"
```

Comment:
Could you maybe explain what this statement does? (And I think some reformatting wouldn't hurt here)